pkgs/build-support/build-fhs-userenv/chrootenv.c

```diff
@@ -0,0 +1,181 @@
+   {"/dev", "dev"},
+   {"/home", "home"}};
+
+void bind(struct bind *bind)
+{
+  DIR *src = opendir(bind->from);
+
+  if (src) {
+    if (closedir(src) == -1)
+      errorf(EX_IOERR, "closedir");
+
+    if (mkdir(bind->to, 0755) == -1)
+      errorf(EX_IOERR, "mkdir");
+
+    if (mount(bind->from, bind->to, "bind", MS_BIND | MS_REC, NULL) == -1)
+      errorf(EX_OSERR, "mount");
+
+  } else {
+    // https://github.com/NixOS/nixpkgs/issues/31104
+    if (errno != ENOENT)
+      errorf(EX_OSERR, "opendir");
+  }
+}
+
+void spitf(char *path, char *fmt, ...)
+{
+  va_list args;
+  va_start(args, fmt);
+
+  FILE *f = fopen(path, "w");
+
+  if (f == NULL)
+    errorf(EX_IOERR, "spitf(%s): fopen", path);
+
+  if (vfprintf(f, fmt, args) == -1)
+    errorf(EX_IOERR, "spitf(%s): vfprintf", path);
+
+  fclose(f); // ignore for now, fails on /proc/self/{g,u}id_map
+}
+
+int nftw_rm(const char *path, const struct stat *sb, int type, struct FTW *ftw)
+{
+  if(remove(path) == -1)
+    errorf(EX_IOERR, "nftw_rm");
+}
+
+#define LEN(x) sizeof(x) / sizeof(*x)
+
+int main(int argc, char *argv[])
+{
+  if (argc < 2) {
+    fprintf(stderr, "Usage: %s command [arguments...]\n"
+                    "Requires Linux kernel >= 3.19 with CONFIG_USER_NS.\n", argv[0]);
+    exit(EX_USAGE);
+  }
+
+  char *root = mkdtemp(strdup("/tmp/chrootenvXXXXXX"));
+
+  if (root == NULL)
+    errorf(EX_IOERR, "mkdtemp");
+
+  // Don't make root private so that privilege drops inside chroot are possible:
+  if (chmod(root, 0755) == -1)
+    errorf(EX_IOERR, "chmod");
+
+  pid_t cpid = fork();
+
+  if (cpid == -1)
+    errorf(EX_OSERR, "fork");
+
+  if (cpid == 0) {
+    uid_t uid = getuid();
+    gid_t gid = getgid();
+
+    if (uid == -1)
+      errorf(EX_OSERR, "getuid");
+
+    if (gid == -1)
+      errorf(EX_OSERR, "getgid");
+
+    // If we are root, no need to create new user namespace.
+    if (uid == 0) {
+      unshare(CLONE_NEWNS);
+      // Mark all mounted filesystems as slave so changes
+      // don't propagate to the parent mount namespace.
+      mount(NULL, "/", NULL, MS_REC | MS_SLAVE, NULL);
+    } else {
+      // Create new mount and user namespaces. CLONE_NEWUSER
+      // requires a program to be non-threaded.
+      unshare(CLONE_NEWNS | CLONE_NEWUSER);
+
+      // Map users and groups to the parent namespace.
+      // setgroups is only available since Linux 3.19:
+      spitf("/proc/self/setgroups", "deny");
+
+      spitf("/proc/self/uid_map", "%d %d 1", uid, gid);
+      spitf("/proc/self/gid_map", "%d %d 1", uid, gid);
+    }
+
+    if (chdir(root) == -1)
+      errorf(EX_IOERR, "chdir");
+
+    for (size_t i = 0; i < LEN(binds); i++)
+      bind(&binds[i]);
+
+    if(chroot(root) == -1)
+      errorf(EX_OSERR, "chroot");
+
+    if(chdir("/") == -1)
+      errorf(EX_OSERR, "chdir");
+
+    argv++;
+
+    if (execvpe(argv[0], argv, env_build(env_whitelist, LEN(env_whitelist))) == -1)
+      errorf(EX_OSERR, "execvpe");
+  }
+
+  while (wait(NULL) > 0);
```

Comment:
I would only wait for our child process here and exit with the same exit status:

I have here some code, I adapted from `nsenter`:
https://github.com/Mic92/nsattach/blob/master/nsattach.c#L376

Reply:
Thanks! Now doing that.

Reply:
Thanks, now doing that.

I've added signal handling code but no `SIGSTOP` handling. I don't think that is the right thing to do here because it means that sending `SIGCONT` to child process without repeating it to parent would result in parent being suspended indefinitely.